`.map()` on a streaming IterableDataset silently skips examples after load_state_dict

### Describe the bug

After `ds.map(fn)` on a streaming IterableDataset, saving state_dict() and loading it into a fresh dataset does not resume at the correct position. It jumps forward by up to one Arrow read-chunk (~1000 rows).

### Steps to reproduce the bug

```python
from datasets import Dataset

ds1 = Dataset.from_dict({"id": list(range(5000))}).to_iterable_dataset(num_shards=1).map(
    lambda x: {**x, "tag": "X"}
)
ds2 = Dataset.from_dict({"id": list(range(5000))}).to_iterable_dataset(num_shards=1).map(
    lambda x: {**x, "tag": "X"}
)

it = iter(ds1)
for _ in range(200):
    next(it)
state = ds1.state_dict()
expected = [next(it)["id"] for _ in range(5)]

ds2.load_state_dict(state)
it2 = iter(ds2)
actual = [next(it2)["id"] for _ in range(5)]

print(f"expected={expected}")
print(f"actual  ={actual}")
assert expected == actual, f"resume skipped forward by {actual[0] - expected[0]} rows"
```

```
expected=[200, 201, 202, 203, 204]
actual  =[1000, 1001, 1002, 1003, 1004]
AssertionError: resume skipped forward by 800 rows
```

### Expected behavior

Expected behaviour:
```
expected=[200, 201, 202, 203, 204]
actual  =[200, 201, 202, 203, 204]
```

### Environment info

```
- `datasets` version: 4.8.4
- Platform: Linux-6.8.0-85-generic-x86_64-with-glibc2.39
- Python version: 3.12.12
- `huggingface_hub` version: 1.11.0
- PyArrow version: 23.0.1
- Pandas version: 3.0.2
- `fsspec` version: 2026.2.0
```

--- a/src/datasets/iterable_dataset.py
+++ b/src/datasets/iterable_dataset.py
@@ -2222,7 +2222,7 @@ class FormattedExamplesIterable(_BaseExamplesIterable):
         # It's ok to use _iter_arrow here without fancy state_dict logic since it's
         # used with RebatchedArrowExamplesIterable with the right batch_size to
         # never lose examples
-        if self.ex_iterable.iter_arrow and not self.force_convert_to_python:
+        if self.ex_iterable.iter_arrow:
             # feature casting (inc column addition) handled within self._iter_arrow()
             for key, pa_table in self._iter_arrow():
                 batch = formatter.format_batch(pa_table)
@@ -3428,10 +3428,11 @@ class IterableDataset(DatasetInfoMixin):
                 force_convert_to_arrow=True,
             )
         else:
-            if self._formatting and self._ex_iterable.iter_arrow:
-                ex_iterable = RebatchedArrowExamplesIterable(
-                    self._ex_iterable, batch_size=batch_size if batched else 1, drop_last_batch=drop_last_batch
-                )
+            if self._ex_iterable.iter_arrow:
+                if self._formatting or input_features:
+                    ex_iterable = RebatchedArrowExamplesIterable(
+                        self._ex_iterable, batch_size=batch_size if batched else 1, drop_last_batch=drop_last_batch
+                    )
             if self._formatting or input_features:
                 # apply formatting after iter_arrow to avoid re-encoding the examples
                 ex_iterable = FormattedExamplesIterable(

--- a/src/datasets/iterable_dataset.py
+++ b/src/datasets/iterable_dataset.py
@@ -2222,7 +2222,7 @@ class FormattedExamplesIterable(_BaseExamplesIterable):
         # It's ok to use _iter_arrow here without fancy state_dict logic since it's
         # used with RebatchedArrowExamplesIterable with the right batch_size to
         # never lose examples
-        if self.ex_iterable.iter_arrow:
+        if self.ex_iterable.iter_arrow and not self.force_convert_to_python:
             # feature casting (inc column addition) handled within self._iter_arrow()
             for key, pa_table in self._iter_arrow():
                 batch = formatter.format_batch(pa_table)

